app/wasm/message_plugin.go

```diff
@@ -0,0 +1,201 @@
+	}
+	err = b.SendCoinsFromModuleToAccount(ctx, gammtypes.ModuleName, rcpt, coins)
+	if err != nil {
+		return sdkerrors.Wrap(err, "sending newly minted coins from message")
+	}
+	return nil
+}
+
+func (m *CustomMessenger) swapTokens(ctx sdk.Context, contractAddr sdk.AccAddress, swap *wasmbindings.SwapMsg) ([]sdk.Event, [][]byte, error) {
+	_, err := PerformSwap(m.gammKeeper, ctx, contractAddr, swap)
+	if err != nil {
+		return nil, nil, sdkerrors.Wrap(err, "perform swap")
+	}
+	return nil, nil, nil
+}
+
+// PerformSwap can be used both for the real swap, and the EstimateSwap query
+func PerformSwap(keeper *gammkeeper.Keeper, ctx sdk.Context, contractAddr sdk.AccAddress, swap *wasmbindings.SwapMsg) (*wasmbindings.SwapAmount, error) {
+	if swap == nil {
+		return nil, wasmvmtypes.InvalidRequest{Err: "gamm perform swap null swap"}
+	}
+	if swap.Amount.ExactIn != nil {
+		routes := []gammtypes.SwapAmountInRoute{{
+			PoolId:        swap.First.PoolId,
+			TokenOutDenom: swap.First.DenomOut,
+		}}
+		for _, step := range swap.Route {
+			routes = append(routes, gammtypes.SwapAmountInRoute{
+				PoolId:        step.PoolId,
+				TokenOutDenom: step.DenomOut,
+			})
+		}
+		if swap.Amount.ExactIn.Input.IsNegative() {
+			return nil, wasmvmtypes.InvalidRequest{Err: "gamm perform swap negative amount in"}
+		}
+		tokenIn := sdk.Coin{
+			Denom:  swap.First.DenomIn,
+			Amount: swap.Amount.ExactIn.Input,
+		}
+		tokenOutMinAmount := swap.Amount.ExactIn.MinOutput
+		estimatedAmount, err := keeper.MultihopSwapExactAmountIn(ctx, contractAddr, routes, tokenIn, tokenOutMinAmount)
+		if err != nil {
+			return nil, sdkerrors.Wrap(err, "gamm perform swap exact amount in")
+		}
+		return &wasmbindings.SwapAmount{Out: &estimatedAmount}, nil
+	} else if swap.Amount.ExactOut != nil {
+		routes := []gammtypes.SwapAmountOutRoute{{
+			PoolId:       swap.First.PoolId,
+			TokenInDenom: swap.First.DenomIn,
+		}}
+		output := swap.First.DenomOut
+		for _, step := range swap.Route {
+			routes = append(routes, gammtypes.SwapAmountOutRoute{
+				PoolId:       step.PoolId,
+				TokenInDenom: output,
+			})
+			output = step.DenomOut
+		}
+		tokenInMaxAmount := swap.Amount.ExactOut.MaxInput
+		if swap.Amount.ExactOut.Output.IsNegative() {
+			return nil, wasmvmtypes.InvalidRequest{Err: "gamm perform swap negative amount out"}
+		}
+		tokenOut := sdk.Coin{
+			Denom:  output,
+			Amount: swap.Amount.ExactOut.Output,
+		}
+		estimatedAmount, err := keeper.MultihopSwapExactAmountOut(ctx, contractAddr, routes, tokenInMaxAmount, tokenOut)
+		if err != nil {
+			return nil, sdkerrors.Wrap(err, "gamm perform swap exact amount out")
+		}
+		return &wasmbindings.SwapAmount{In: &estimatedAmount}, nil
+	} else {
+		return nil, wasmvmtypes.UnsupportedRequest{Kind: "must support either Swap.ExactIn or Swap.ExactOut"}
+	}
+}
+
+// GetFullDenom is a function, not method, so the message_plugin can use it
+func GetFullDenom(contract string, subDenom string) (string, error) {
```

Comment:
This is where we build the custom denom if the contract mints a coin.
I assume this is enough, but if you need metadata registration or other such, please let us know.